client/src/components/JurisdictionAdmin/BatchRoundDataEntry/BatchDetails.tsx

```diff
@@ -0,0 +1,472 @@
+      )}
+
+      <BatchResultTallySheet
+        addSheet={addSheet}
+        areResultsFinalized={areResultsFinalized}
+        batch={batch}
+        contest={contest}
+        deleteSheet={deleteSheet}
+        isEditing={isEditing}
+        key={selectedTabId}
+        selectedTabId={selectedTabId}
+        setAreChangesUnsaved={setAreChangesUnsaved}
+        setIsEditing={setIsEditing}
+        sheets={sheets}
+        updateSheet={updateCurrentSheet}
+      />
+    </Detail>
+  )
+}
+
+interface IBatchResultTallySheetProps {
+  addSheet: () => Promise<void>
+  areResultsFinalized: boolean
+  batch: IBatch
+  contest: IContest
+  deleteSheet: () => Promise<void>
+  isEditing: boolean
+  selectedTabId: string
+  setAreChangesUnsaved: (areChangesUnsaved: boolean) => void
+  setIsEditing: (isEditing: boolean) => void
+  sheets: IBatchResultTallySheetWithId[]
+  updateSheet: (updatedSheet: IBatchResultTallySheet) => Promise<void>
+
+  // Require a key to ensure that the state within this component resets when a different sheet is
+  // selected
+  key: string // eslint-disable-line react/no-unused-prop-types
+}
+
+const BatchResultTallySheet: React.FC<IBatchResultTallySheetProps> = ({
+  addSheet,
+  areResultsFinalized,
+  batch,
+  contest,
+  deleteSheet,
+  isEditing,
+  selectedTabId,
+  setAreChangesUnsaved,
+  setIsEditing,
+  sheets,
+  updateSheet,
+}) => {
+  const tabs = tabsFromSheets(sheets)
+  const isTotalsSheet = selectedTabId === VOTE_TOTALS_TAB_ID
+  const selectedSheet = isTotalsSheet
+    ? undefined
+    : sheets.find(sheet => sheet.id === selectedTabId)
+
+  const formMethods = useForm<IBatchResultTallySheet>({
+    defaultValues: selectedSheet,
+  })
+  const { errors, formState, handleSubmit, register, reset } = formMethods
+  // Important gotcha! You have to access properties on the formState to subscribe to it:
+  // https://github.com/react-hook-form/react-hook-form/issues/9002
+  const { isSubmitting, isDirty } = formState
+
+  // Communicate up to the parent whether or not there are unsaved changes
+  useEffect(() => {
+    setAreChangesUnsaved(isDirty)
+  }, [isDirty, setAreChangesUnsaved])
+
+  const discardChanges = () => {
+    reset(selectedSheet)
+    setIsEditing(false)
+  }
+
+  const onValidSubmit: SubmitHandler<IBatchResultTallySheet> = async sheet => {
+    try {
+      await updateSheet(sheet)
+    } catch {
+      // Errors are automatically toasted by the queryClient
+      return
+    }
+    // Reset the form's isDirty value back to false
+    reset(sheet)
+    setIsEditing(false)
+  }
+
+  return (
+    <>
+      {isEditing && sheets.length > 1 ? (
+        // A special tab bar with a sheet name form input replacing the selected tab, rendered here
+        // instead of in the parent component so that we can create a separate react-hook-form form
+        // per sheet (sharing the form across sheets requires diligent resetting)
+        <Tabs id={batch.name} selectedTabId={selectedTabId}>
```

Comment:
An interesting tradeoff here. I could've either gone with one react-hook-form form shared across sheets, which would've allowed for one tab bar, or two tab bars, which allows for a separate react-hook-form form per sheet (making state management bugs due to incorrect resetting less likely).

The latter felt more important to prioritize. One remaining negative side effect of the two tab bars is that, after you finish editing a sheet (and we return to a different tab bar under the hood), we reanimate the selection of the already selected tab:

![199039307-625a4aed-84c6-4a97-b911-a35698608564](https://user-images.githubusercontent.com/12616928/199044405-2be617fc-8e99-40a0-a7b3-ffcd70c66eec.gif)

Still thinking through workarounds, one of which is just disabling tab bar animation altogether by setting the `animate` prop to false. Already tried toggling the `animate` prop intelligently to just prevent the initial animation and not subsequent ones, but that interfered with the selected tab indicator in other ways.

Reply:
Oof this is tough! I agree with your prioritization, though I do love the animation. Did you try lifting up the react-hook-form and see bugs, or was that just a preventative choice? Maybe we leave it as is, disable animation for now and try to find a workaround as a follow up?

Reply:
I originally did have the react-hook-form up a level, and while I eventually worked through the bugs, the resulting code was complex and felt prone to bugs down the road, which started to defeat one of the key purposes of switching from https://github.com/votingworks/arlo/pull/1644 to this

That said, I came up with a way to toggle animation on and off that works! https://github.com/votingworks/arlo/pull/1661/commits/c844e4946da0f211188156018f722191a173ca76

I also really like the animation 😄